Adding support for partition override as rollback plan

jsjtzyy opened the pull request:
- introduce PartitionOverride into HelixPropertyStore to store partition
sealed state from static clustermap
- add uploadConfig method in HelixBootstrapUpgradeTool to support upload
partition info from json file
- add new clusterMapConfig to enable/disable partition override
- make ClusterManager take proper actions when instanceConfig changes
based on config of partition override
- ./gradlew build && ./gradlew test

jsjtzyy commented on ambry-clustermap/src/main/java/com.github.ambry.clustermap/HelixClusterManager.java:
```diff
@@ -157,6 +171,35 @@ public void run() {
+    };
+    logger.info("Getting ZNRecord from Helix PropertyStore");
+    helixPropertyStore.subscribe(ZNODE_PATH, helixListener);
+    ZNRecord zNRecord = helixPropertyStore.get(ZNODE_PATH, null, AccessOption.PERSISTENT);
```
Yeah, in current PR, we listen to changes of override map but do nothing. The methods here are kept for future use. (Later, we might need to change the content of override map without rebooting.)

jsjtzyy commented:
@cgtz addressed comments, @pnarayanan @vgkholla remind to review.

cgtz commented on ambry-clustermap/src/main/java/com.github.ambry.clustermap/HelixBootstrapUpgradeUtil.java:
```diff
@@ -231,6 +268,57 @@ private HelixBootstrapUpgradeUtil(String hardwareLayoutPath, String partitionLay
+    for (PartitionId partitionId : staticClusterMap.getAllPartitionIds(null)) {
+      String partitionName = partitionId.toPathString();
+      Map<String, String> partitionProperties = new HashMap<>();
+      partitionProperties.put("state", "RW");
```
Make these states into an enum

cgtz commented on ambry-clustermap/src/main/java/com.github.ambry.clustermap/HelixBootstrapUpgradeUtil.java:
```diff
@@ -231,6 +268,57 @@ private HelixBootstrapUpgradeUtil(String hardwareLayoutPath, String partitionLay
+      ZNRecord znRecord = new ZNRecord(ZNRECORD_NAME);
+      znRecord.setMapFields(partitionOverrideInfos);
+      String path = "/" + this.clusterName + PROPERTYSTORE_PATH + "/" + ZNRECORD_NAME;
+      helixPropertyStore.set(path, znRecord, AccessOption.PERSISTENT);
```
Check the return value of `set`. 

Also, you may not need to create the znode ahead of time in `createHelixPropertyStoreForAllDCs()` since the `set` method will take care of that. In fact, if you do this, you would not need `createHelixPropertyStoreForAllDCs()` and you could just instantiate the zk client within `uploadPartitionOverride`

jsjtzyy replied:
sounds reasonable. I will make it more concise.

cgtz commented on ambry-clustermap/src/main/java/com.github.ambry.clustermap/HelixBootstrapUpgradeUtil.java:
```diff
@@ -231,6 +268,57 @@ private HelixBootstrapUpgradeUtil(String hardwareLayoutPath, String partitionLay
+          propertyStoreConfig.zkClientConnectionTimeoutMs, new ZNRecordSerializer());
+      List<String> subscribedPaths = Collections.singletonList(propertyStoreConfig.rootPath);
+      HelixPropertyStore<ZNRecord> helixPropertyStore =
+          new ZkHelixPropertyStore<>(new ZkBaseDataAccessor<>(zkClient), propertyStoreConfig.rootPath, subscribedPaths);
```
In this situation, you can pass in `null` to `subscribedPaths` since you are not using the listener functionality.

cgtz commented on ambry-tools/src/main/java/com.github.ambry/store/DiskReformatter.java:
```diff
@@ -158,10 +159,11 @@ public static void main(String[] args) throws Exception {
     ClusterAgentsFactory clusterAgentsFactory =
-        Utils.getObj(clusterMapConfig.clusterMapClusterAgentsFactory, clusterMapConfig, config.hardwareLayoutFilePath,
-            config.partitionLayoutFilePath);
+        Utils.getObj(clusterMapConfig.clusterMapClusterAgentsFactory, clusterMapConfig, helixPropertyStoreConfig,
```
I think this change should be reverted since you did not end up changing the `ClusterAgentsFactory` constructor arguments

cgtz commented on ambry-tools/src/main/java/com.github.ambry/store/StoreCopier.java:
```diff
@@ -172,9 +173,10 @@ public static void main(String[] args) throws Exception {
     ClusterAgentsFactory clusterAgentsFactory =
-        Utils.getObj(clusterMapConfig.clusterMapClusterAgentsFactory, clusterMapConfig, config.hardwareLayoutFilePath,
-            config.partitionLayoutFilePath);
+        Utils.getObj(clusterMapConfig.clusterMapClusterAgentsFactory, clusterMapConfig, helixPropertyStoreConfig,
```
This change should be reverted too.

cgtz commented on ambry-tools/src/integration-test/java/com.github.ambry/clustermap/HelixBootstrapUpgradeToolTest.java:
```diff
@@ -209,6 +232,44 @@ private void writeBootstrapOrUpgrade(long expectedResourceCount, boolean forceRe
+      ZkClient zkClient = new ZkClient("localhost:" + zkInfo.getPort(), propertyStoreConfig.zkClientSessionTimeoutMs,
+          propertyStoreConfig.zkClientConnectionTimeoutMs, new ZNRecordSerializer());
+      List<String> subscribedPaths = Collections.singletonList(propertyStoreConfig.rootPath);
+      HelixPropertyStore<ZNRecord> propertyStore =
```
We write very similar code to set up `HelixPropertyStore` in many places: `HelixBootstrapUpgradeTool`, `HelixPropertyStoreFactory`, `HelixNotifier`, etc. I think a utility method to instantiate `HelixPropertyStore` could help.

jsjtzyy replied:
true. Let me make relevant changes.

cgtz commented on ambry-tools/src/integration-test/java/com.github.ambry/clustermap/HelixBootstrapUpgradeToolTest.java:
```diff
@@ -48,6 +61,10 @@
   private static final String CLUSTER_NAME_IN_STATIC_CLUSTER_MAP = "ToolTestStatic";
   private static final String CLUSTER_NAME_PREFIX = "Ambry-";
+  private static final String ROOT_PATH = "/" + CLUSTER_NAME_PREFIX + CLUSTER_NAME_IN_STATIC_CLUSTER_MAP;
+  private static final String ZNRECORD_NAME = "PartitionOverride";
```
could you refer to the constants in `HelixBootstrapUpgradeUtil`?

jsjtzyy commented on ambry-clustermap/src/main/java/com.github.ambry.clustermap/HelixClusterManager.java:
```diff
@@ -92,6 +103,10 @@
+      // Make sure the HelixManager of local datacenter gets connected first and partitionOverrideInfoMap use PropertyStore
+      // in local DC for initialization.
+      HelixManager localManager =
+          initializeHelixManagerAndPropertyStoreInLocalDC(dataCenterToZkAddress, instanceName, helixFactory);
```
The intention here is to make sure clusterManager connects to HelixManager in local datacenter first. (There are two benefits: 1. the `PropertyStore` might vary among colos later, clusterManager should use `PropertyStore` in local DC.  2. There is race condition if clusterManager concurrently connects to 4 HelixManager in all DCs)

jsjtzyy replied:
The latest code changes and deployment plan are updated in Design doc. Also the race condition is discussed there.

jsjtzyy commented on ambry-clustermap/src/main/java/com.github.ambry.clustermap/HelixClusterManager.java:
```diff
@@ -102,11 +117,16 @@
-              manager.connect();
-              logger.info("Established connection to Helix manager at {}", zkConnectStr);
+              HelixManager manager;
+              if (!dcName.equals(clusterMapConfig.clusterMapDatacenterName)) {
```
skip the HelixManager for local DC. Not sure how much delay it will bring. (Basically we change 4 concurrent threads to 1 thread first and 3 concurrent threads)

pnarayanan replied:
shouldn't be a problem, though is there a reason to not initialize overrides in parallel?

jsjtzyy replied:
for race condition concerns, will talk to you offline.

vgkholla commented on ambry-clustermap/src/main/java/com.github.ambry.clustermap/HelixClusterManager.java:
```diff
@@ -157,6 +177,58 @@ public void run() {
+      manager.connect();
+      logger.info("Established connection to Helix manager in local datacenter at {}", zkConnectStr);
+    } catch (Exception e) {
+      initializationException.compareAndSet(null, e);
```
should this be an immediate failure? This is not being done in its own thread - it's in the main thread, right?

jsjtzyy commented on ambry-api/src/main/java/com.github.ambry/config/ClusterMapConfig.java:
```diff
@@ -138,6 +138,13 @@
   public final String clusterMapDefaultPartitionClass;
 
+  /**
+   * Indicate if cluster manager enables override on dynamic sealed list.
```
done.

jsjtzyy commented on ambry-api/src/main/java/com.github.ambry/config/ClusterMapConfig.java:
```diff
@@ -138,6 +138,13 @@
+  /**
+   * Indicate if cluster manager enables override on dynamic sealed list.
+   */
+  @Config("clustermap.enable.override")
```
How about `clustermap.enable.partition.override`?

jsjtzyy commented on ambry-clustermap/src/main/java/com.github.ambry.clustermap/ClusterMapUtils.java:
```diff
@@ -122,6 +132,21 @@ public static String getInstanceName(String host, Integer port) {
+   * @return the instance of {@link HelixPropertyStore}.
+   */
+  public static HelixPropertyStore<ZNRecord> createHelixPropertyStore(String zkServers,
+      HelixPropertyStoreConfig propertyStoreConfig, List<String> subscribedPaths) {
```
This is because the `zkServer` (a.k.a `zkClientConnectString` in config) can be different from that in `HelixPropertyStoreConfig`, especially in some unit tests. That's why I didn't pick zk path from `HelixPropertyStoreConfig`. Since the `zkClientConnectString` is a general config which is better to be in `HelixPropertyStoreConfig`, I prefer to pass the args separately.

jsjtzyy replied:
No, I are not. The property store for account service could be explicitly specified by `AccountServiceConfig.zkClientConnectStr`. They are still decoupled.

jsjtzyy commented on ambry-clustermap/src/main/java/com.github.ambry.clustermap/ClusterMapUtils.java:
```diff
@@ -122,6 +132,21 @@ public static String getInstanceName(String host, Integer port) {
+   * @param subscribedPaths a list of paths to which the PropertyStore subscribes.
+   * @return the instance of {@link HelixPropertyStore}.
+   */
+  public static HelixPropertyStore<ZNRecord> createHelixPropertyStore(String zkServers,
```
Unit test added. Might take a little more time for testing.

vgkholla commented on ambry-clustermap/src/main/java/com.github.ambry.clustermap/ClusterMapUtils.java:
```diff
@@ -122,6 +135,25 @@ public static String getInstanceName(String host, Integer port) {
+   * @throws IllegalArgumentException
+   */
+  public static HelixPropertyStore<ZNRecord> createHelixPropertyStore(String zkServers,
+      HelixPropertyStoreConfig propertyStoreConfig, List<String> subscribedPaths) throws IllegalArgumentException {
```
you need not explicitly declare `throws IllegalArgumentException` since it is a `RuntimeException`

vgkholla commented on ambry-clustermap/src/main/java/com.github.ambry.clustermap/HelixBootstrapUpgradeUtil.java:
```diff
@@ -216,9 +246,10 @@ private HelixBootstrapUpgradeUtil(String hardwareLayoutPath, String partitionLay
-    info("Associating static Ambry cluster \"" + clusterNameInStaticClusterMap + "\" with cluster\"" + clusterName
-        + "\" in Helix");
+    clusterName = clusterNamePrefix + clusterNameInStaticClusterMap;
+    System.out.println(
```
why change this from `info` to `sout`?

jsjtzyy replied:
oh, my mistake. I should change it back to `info` after debugging.

vgkholla commented on ambry-clustermap/src/main/java/com.github.ambry.clustermap/HelixClusterManager.java:
```diff
@@ -312,7 +384,7 @@ public void onInstanceConfigChange(List<InstanceConfig> configs, NotificationCon
           instanceConfigInitialized.set(true);
-        } else {
+        } else if (!clusterMapConfig.clusterMapEnablePartitionOverride) {
           updateSealedStateOfReplicas(configs);
```
your other PR made a change to this function (to account for stopped replicas too) - does this change in the `if` still hold? (I think it should)

jsjtzyy replied:
hmm, i think here might be a problem. If `clusterMapEnablePartitionOverride` is enabled, looks like we cannot dynamically update the `STOPPED` state of replica. What do you think? (I didn't consider this case in that PR)

jsjtzyy commented:
address comments from @vgkholla  (./gradlew build && ./gradlew test succeeded)

vgkholla commented on ambry-clustermap/src/test/java/com.github.ambry.clustermap/HelixClusterManagerTest.java:
```diff
@@ -152,8 +178,9 @@ public HelixClusterManagerTest(boolean useComposite) throws Exception {
     props.setProperty("clustermap.cluster.name", clusterNamePrefixInHelix + clusterNameStatic);
     props.setProperty("clustermap.datacenter.name", localDc);
     props.setProperty("clustermap.dcs.zk.connect.strings", zkJson.toString(2));
+    props.setProperty("clustermap.enable.partition.override", this.overrideEnabled ? "true" : "false");
```
minor: don't need this

vgkholla commented on ambry-clustermap/src/test/java/com.github.ambry.clustermap/HelixClusterManagerTest.java:
```diff
@@ -428,12 +493,46 @@ public void sealedReplicaChangeTest() {
+   */
+  @Test
+  public void clusterMapOverrideEnabledTest() throws Exception {
+    if (overrideEnabled) {
```
is there a test that makes sure overrides are not used (even if they exist) if the config is not enabled?

vgkholla commented on ambry-clustermap/src/test/java/com.github.ambry.clustermap/HelixClusterManagerTest.java:
```diff
@@ -428,12 +493,46 @@ public void sealedReplicaChangeTest() {
+      // Verify clustermap uses partition override for initialization
+      Set<String> writableInOverrideMap = new HashSet<>();
+      for (Map.Entry<String, Map<String, String>> entry : partitionOverrideMap.entrySet()) {
+        if (entry.getValue().get(ClusterMapUtils.PARTITION_STATE).equals(ClusterMapUtils.READ_WRITE_STR)) {
```
is there a test that makes sure that overrides take effect?

vgkholla commented on ambry-clustermap/src/test/java/com.github.ambry.clustermap/HelixClusterManagerTest.java:
```diff
@@ -178,7 +178,7 @@ public HelixClusterManagerTest(boolean useComposite, boolean overrideEnabled) th
     props.setProperty("clustermap.datacenter.name", localDc);
     props.setProperty("clustermap.dcs.zk.connect.strings", zkJson.toString(2));
-    props.setProperty("clustermap.enable.partition.override", this.overrideEnabled ? "true" : "false");
+    props.setProperty("clustermap.enable.partition.override", overrideEnabled ? "true" : "false");
```
minor: can do `Boolean.toString(overrideEnabled)` instead

pnarayanan commented on ambry-clustermap/src/main/java/com.github.ambry.clustermap/HelixBootstrapUpgradeUtil.java:
```diff
@@ -140,6 +144,32 @@ static void bootstrapOrUpgrade(String hardwareLayoutPath, String partitionLayout
+            maxPartitionsInOneResource, false, false, helixAdminFactory);
+    Map<String, Map<String, String>> partitionOverrideInfos =
+        clusterMapToHelixMapper.generatePartitionOverrideFromAllDCs();
+    info("Uploading partition override to HelixPropertyStore based on static clustermap.");
```
we should stop using the term "static clustermap" especially for things that are newly introduced.

jsjtzyy replied:
Right, I will change it to avoid using term `static clustermap`

pnarayanan commented on ambry-clustermap/src/main/java/com.github.ambry.clustermap/HelixBootstrapUpgradeUtil.java:
```diff
@@ -231,6 +261,42 @@ private HelixBootstrapUpgradeUtil(String hardwareLayoutPath, String partitionLay
+      HelixPropertyStore<ZNRecord> helixPropertyStore =
+          ClusterMapUtils.createHelixPropertyStore(entry.getValue().getZkConnectStr(), propertyStoreConfig, null);
+      ZNRecord znRecord = new ZNRecord(ClusterMapUtils.ZNODE_NAME);
+      znRecord.setMapFields(partitionOverrideInfos);
```
One problem with the map fields is that you can only have one kind of mapping on the znode. Basically mapping on this znode can only be used for partition override.

jsjtzyy replied:
Currently, it is not a serious problem because the znode is purposely created for partition override map. If there is another map later, we could place it in another znode.

pnarayanan commented on ambry-clustermap/src/main/java/com.github.ambry.clustermap/HelixBootstrapUpgradeUtil.java:
```diff
@@ -231,6 +261,42 @@ private HelixBootstrapUpgradeUtil(String hardwareLayoutPath, String partitionLay
+    storeProps.setProperty("helix.property.store.root.path", "/" + clusterName);
+    HelixPropertyStoreConfig propertyStoreConfig = new HelixPropertyStoreConfig(new VerifiableProperties(storeProps));
+    info("Setting partition override for all datacenters.");
+    for (Map.Entry<String, ClusterMapUtils.DcZkInfo> entry : dataCenterToZkAddress.entrySet()) {
```
So the same information is going to be duplicated in every data center? Do you think that is the right thing to do?

jsjtzyy replied:
The ClusterManager on server only talks to zookeeper in local datacenter to get the override info. Hence, it's necessary to send the override info to all DCs. Also, in the future, the override info may vary by DCs, we should upload the specific override info to partitular DC separately.

pnarayanan commented on ambry-clustermap/src/main/java/com.github.ambry.clustermap/HelixClusterManager.java:
```diff
@@ -102,11 +114,16 @@
-              manager.connect();
-              logger.info("Established connection to Helix manager at {}", zkConnectStr);
+              HelixManager manager;
+              if (!dcName.equals(clusterMapConfig.clusterMapDatacenterName)) {
```
I generally prefer `if (x) then y else z`  vs. `if (!x) then z else y`. Not a big deal here, but in a more general setting, you might end up doing double negation in your mind for the else case which can be annoying.

jsjtzyy replied:
sounds reasonable. Will take your advice.

jsjtzyy commented on ambry-account/src/main/java/com/github/ambry/account/HelixAccountServiceFactory.java:
```diff
@@ -77,10 +74,8 @@ public AccountService getAccountService() {
-          storeConfig.zkClientConnectionTimeoutMs, new ZNRecordSerializer());
       HelixPropertyStore<ZNRecord> helixStore =
-          new ZkHelixPropertyStore<>(new ZkBaseDataAccessor<>(zkClient), storeConfig.rootPath, null);
+          ClusterMapUtils.createHelixPropertyStore(accountServiceConfig.zkClientConnectString, storeConfig, null);
```
So you suggest moving `createHelixPropertyStore` to another module?

vgkholla replied:
The function probably needs a better home but other than this particular static function use (which is unrelated to the actual clustermap), account and clustermap are still decoupled. Any suggestions on where this function can go? `ambry-commons` in a `CommonUtils` class? or in `ambry-api`?

jsjtzyy replied:
Let me move it into `ambry-commons` first.

jsjtzyy commented:
address comments from @pnarayanan ,  ./gradlew build && ./gradlew test succeeded. @cgtz remind to review.

jsjtzyy commented:
Fixed the conflicts after xid code was brought in. @pnarayanan @cgtz  please take a look.

cgtz commented on ambry-clustermap/src/main/java/com.github.ambry.clustermap/HelixBootstrapUpgradeUtil.java:
```diff
@@ -231,6 +262,42 @@ private HelixBootstrapUpgradeUtil(String hardwareLayoutPath, String partitionLay
+    for (Map.Entry<String, ClusterMapUtils.DcZkInfo> entry : dataCenterToZkAddress.entrySet()) {
+      HelixPropertyStore<ZNRecord> helixPropertyStore =
+          CommonUtils.createHelixPropertyStore(entry.getValue().getZkConnectStr(), propertyStoreConfig, null);
+      ZNRecord znRecord = new ZNRecord(ClusterMapUtils.ZNODE_NAME);
```
actually can those constants in CommonUtils be removed?

cgtz commented on ambry-clustermap/src/main/java/com.github.ambry.clustermap/HelixClusterManager.java:
```diff
@@ -159,6 +176,61 @@ public void run() {
+      HelixPropertyListener helixListener = new HelixPropertyListener() {
+        @Override
+        public void onDataChange(String path) {
+          logger.info("Message is changed for topic {} at path {}", ClusterMapUtils.ZNODE_NAME, path);
```
won't the path already include the znode name?

jsjtzyy replied:
fixed

cgtz commented on ambry-clustermap/src/main/java/com.github.ambry.clustermap/HelixClusterManager.java:
```diff
@@ -387,8 +460,15 @@ private void updateStateOfReplicas(List<InstanceConfig> instanceConfigs) {
-                  replica.setSealedState(sealedReplicas.contains(replica.getPartitionId().toPathString()));
-                  replica.setStoppedState(stoppedReplicas.contains(replica.getPartitionId().toPathString()));
+                  if (clusterMapConfig.clusterMapEnablePartitionOverride && partitionOverrideInfoMap.containsKey(
+                      replica.getPartitionId().toPathString())) {
```
minor nit: construct this path string once and save it to a local variable

jsjtzyy replied:
Fixed. Thanks.

cgtz commented on ambry-clustermap/src/main/java/com.github.ambry.clustermap/HelixClusterManager.java:
```diff
@@ -494,9 +575,18 @@ private void initializeDisksAndReplicasOnNode(AmbryDataNode datanode, InstanceCo
             // Create replica associated with this node.
+            boolean isSealed;
+            if (clusterMapConfig.clusterMapEnablePartitionOverride && partitionOverrideInfoMap.containsKey(
+                partitionName)) {
```
just want to make sure that partitionName equals the result of toPathString in this case

jsjtzyy replied:
`"2443:412316860416:max-replicas-all-datacenters,2533:412316860416:max-replicas-all-datacenters,` is an example in InstanceConfig. The `partitionName` here is `2443`, `2553` etc, which should equal to the result of `toPathString`.

cgtz commented on ambry-clustermap/src/test/java/com.github.ambry.clustermap/HelixClusterManagerTest.java:
```diff
@@ -186,6 +215,9 @@ public void after() {
    */
   @Test
   public void badInstantiationTest() throws Exception {
+    if (overrideEnabled) {
```
use `assumeTrue(!overrideEnabled)` here.

jsjtzyy replied:
Good point. Take you suggestion, thanks!

cgtz commented on ambry-clustermap/src/test/java/com.github.ambry.clustermap/HelixClusterManagerTest.java:
```diff
@@ -222,12 +254,44 @@ public void badInstantiationTest() throws Exception {
+   */
+  @Test
+  public void emptyPartitionOverrideTest() throws Exception {
+    if (overrideEnabled) {
```
can use assumeTrue here and in a couple other places as well

jsjtzyy replied:
done

pnarayanan commented on ambry-clustermap/src/main/java/com.github.ambry.clustermap/ClusterMapUtils.java:
```diff
@@ -46,13 +54,18 @@
   static final String SEALED_STR = "SEALED";
   static final String STOPPED_REPLICAS_STR = "STOPPED";
   static final String AVAILABLE_STR = "AVAILABLE";
+  static final String READ_ONLY_STR = "RO";
```
`READ_ONLY_STR` and `SEALED_STR` mean the same thing?

jsjtzyy replied:
The `SEALED_STR` is used for sealed list in InstanceConfig, `READ_ONLY_STR` and `READ_WRITE_STR` are used for the partition override infos. Although they have the same meaning, they are used in different places. For partition override map, it's better to make string concise and same length with `READ_WRITE_STR`. 
```"0": {
            "state": "RO"
        },
        "1": {
            "state": "RW"
        },```

pnarayanan commented on ambry-clustermap/src/main/java/com.github.ambry.clustermap/HelixBootstrapUpgradeUtil.java:
```diff
@@ -140,6 +145,32 @@ static void bootstrapOrUpgrade(String hardwareLayoutPath, String partitionLayout
+   */
+  static void uploadClusterConfigs(String hardwareLayoutPath, String partitionLayoutPath, String zkLayoutPath,
+      String clusterNamePrefix, int maxPartitionsInOneResource, HelixAdminFactory helixAdminFactory) throws Exception {
+
```
whitespace

jsjtzyy replied:
fixed.

jsjtzyy commented on ambry-clustermap/src/main/java/com.github.ambry.clustermap/HelixBootstrapUpgradeUtil.java:
```diff
@@ -231,6 +262,42 @@ private HelixBootstrapUpgradeUtil(String hardwareLayoutPath, String partitionLay
   }
 
+  /**
+   * Generate the partition override map containing partition state from all datacenters.
```
done. Will verify it in Perf first.

jsjtzyy commented on ambry-clustermap/src/main/java/com.github.ambry.clustermap/HelixBootstrapUpgradeUtil.java:
```diff
@@ -231,6 +262,42 @@ private HelixBootstrapUpgradeUtil(String hardwareLayoutPath, String partitionLay
+  /**
+   * Uploads the seal state of all partitions in the format of map.
+   */
+  private void uploadPartitionOverride(Map<String, Map<String, String>> partitionOverrideInfos) {
```
done

jsjtzyy commented:
Addressed comments from @cgtz and @pnarayanan. I need to verify uploading/reading partition override map in perf cluster before this PR can be merged. 
./gradlew build && ./gradlew test succeeded.

jsjtzyy commented:
Verified 3 things in Perf cluster:
1. use Helix tool to upload/modify partition override map in `PROPERTYSTORE`;
2. server can get partition override infos from HelixPropertyStore when booting up;
3. server can dynamically receive the notification when data in `PROPERTYSTORE` is changed.

./gradlew build && ./gradlew test succeeded.
